aqt/metadata.py

```diff
@@ -0,0 +1,657 @@
+            _match = pattern.match(name)
+            if not _match:
+                return None, None
+            module_with_arch = _match.group(3)
+            if self.archive_id.is_no_arch() or "." not in module_with_arch:
+                return module_with_arch, None
+            module, arch = module_with_arch.rsplit(".", 1)
+            return module, arch
+
+        rest_of_url = self.archive_id.to_url(
+            qt_version_no_dots=qt_ver_str, file="Updates.xml"
+        )
+        xml = self.fetch_http(rest_of_url)  # raises RequestException
+
+        # We want the names of modules, regardless of architecture:
+        modules = xml_to_modules(
+            xml,
+            predicate=ListCommand._has_nonempty_downloads,
+            keys_to_keep=(),  # Just want names
+        )
+
+        def naive_modules_arches(
+            names: Iterable[str],
+        ) -> Tuple[List[str], List[str]]:
+            modules_and_arches, _modules, arches = set(), set(), set()
+            for name in names:
+                # First term could be a module name or an architecture
+                first_term, arch = to_module_arch(name)
+                if first_term:
+                    modules_and_arches.add(first_term)
+                if arch:
+                    arches.add(arch)
+            for first_term in modules_and_arches:
+                if first_term not in arches:
+                    _modules.add(first_term)
+            return (
+                sorted(_modules),
+                sorted(arches),
+            )
+
+        return naive_modules_arches(modules.keys())
+
+    def describe_filters(self) -> str:
+        if self.filter_minor is None:
+            return str(self.archive_id)
+        return "{} with minor version {}".format(self.archive_id, self.filter_minor)
+
+    def print_suggested_follow_up(self, printer: Callable[[str], None]) -> None:
+        """Makes an informed guess at what the user got wrong, in the event of an error."""
+        base_cmd = "aqt {0.category} {0.host} {0.target}".format(self.archive_id)
+        if self.archive_id.extension:
+            msg = "Please use '{} --extensions <QT_VERSION>' to list valid extensions.".format(
+                base_cmd
+            )
+            printer(msg)
+
+        if self.archive_id.is_tools() and self.request_type == "tool variant names":
+            msg = "Please use '{}' to check what tools are available.".format(base_cmd)
+            printer(msg)
+        elif self.filter_minor is not None:
+            msg = "Please use '{}' to check that versions of {} exist with the minor version '{}'".format(
+                base_cmd, self.archive_id.category, self.filter_minor
+            )
+            printer(msg)
+        elif self.request_type in ("architectures", "modules", "extensions"):
+            msg = "Please use '{}' to show versions of Qt available".format(base_cmd)
+            printer(msg)
```

Comment:
657 lines? You're right, it really did need its own file! Thank you for the refactor!